### DISCUSSION TITLE
WPF Transparent Web Page cannot penetrate mouse events

### DISCUSSION THREAD

**Discussion by qiujinchao** [cefsharp/CefSharp#4052] [General]
I'm using WPF.CefSharp.

I test load a transparent web page, but mouse events cannot penetrate the WPF controls below the transparent web page. 

Is there any other solution to enable mouse events to penetrate transparent web pages.

I read the source and found that e.handle = true, which seems to be the reason why mouse events cannot penetrate.

I tested and found that if the transparent window is above the Winform window in the WindowFormHost, mouse events can penetrate the transparent web page.

**Comment by qiujinchao** [comment]
My demo has two layers of WPF windows, the upper one is a transparent window, which hosts the ChromiumWebBrowser control and loads a transparent web page.

**Comment by amaitland** [comment]
> I test load a transparent web page, but mouse events cannot penetrate the WPF controls below the transparent web page.

Sending mouse events to the browser is the intended behaviour.

What's stopping you from overriding the mouse methods and setting `Handled = false`?

> I tested and found that if the transparent window is above the Winform window in the WindowFormHost, mouse events can penetrate the transparent web page.

`WindowsFormsHost` controls are always on top, so no surprised it behaves differently.

**Reply by qiujinchao** [reply]
My demo has two layers of WPF windows, the upper one is a transparent window, which hosts the ChromiumWebBrowser control and loads a transparent web page，the following window uses WIndowFormHost。

**Reply by qiujinchao** [reply]
e.handle = true is the source code inside CefSharp, can I block the operation of CefSHarp from outside?

**Reply by amaitland** [reply]
> the following window uses WIndowFormHost。

Your going to have `AirSpace` issues. Long as you understand that (nothing to do with `CefSharp`)

> can I block the operation of CefSHarp from outside?

Create a class that inherits from `ChromiumWebBrowser`, override the mouse methods that you need to, override the `Handled` property after you've called the base method.

**Reply by qiujinchao** [reply]
Yes.Because of the airspace, I used two layers of windows. This has nothing to do with CefSharp. I am just surprised that the winform window below the transparent web page can receive mouse events while WPF cannot. I will try to override the mouse events needed, thanks

**Comment by qiujinchao** [comment]
I find that I have a misunderstanding. I've always thought that click events on transparent controls could penetrate to the controls below, but they can't. Only transparent controls on transparent windows can penetrate to the next layer of windows.
